### ISSUE TITLE
Doesn't work that well.

### ISSUE THREAD

**Issue by forkoz** [tarampampam/random-user-agent#36]
## Steps to reproduce:
 1. Enable Extension
 2. Enable "JS" protection
 3. navigate to : https://browserleaks.com/javascript

## Expected behaviour:
Should see platform modified and also the same stuff when you click iframe window.

## Actual behaviour:
See real platform and real UA.

---------

### System information:
**Browser version**:
```
Chromium 61
```

**Comment by neroux** [comment]
#7 and #26

**Comment by forkoz** [comment]
Maybe... clicking over to iframe is way after the page is loaded. This plugin:

https://chrome.google.com/webstore/detail/user-agent-switcher-for-c/djflhoibgkdhkhhcedjiklpkjnoahfmg?hl=en

Changes platform. You can crib that code. The cookie approach seems straight up crazy. The plugin does not handle JS-iframe though either... in fact none of them do and most change the user agent incorrectly in any case. Chrome is without a proper user agent spoofer and that blows my mind.

**Comment by forkoz** [comment]
Also check out chameleon and x-agent extensions.

Changing UA is fighting google it seems: https://bugs.chromium.org/p/chromium/issues/detail?id=395932

**Comment by neroux** [comment]
Thank you, lets please continue in the existing tickets though.

Yes, it seems to be quite tricky to change the user agent throughout the system and unfortunately the cookie approach is not as crazy as you said, it is not without problems either though and still requires some tweaking.

The other navigator properties, the iframe, and the delayed property overload are three distinct problems however, with the former being the easiest to fix.

**Comment by neroux** [comment]
For the time being it seems there is no proper fix for that issue. It has been [known for quite some time](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-extensions/8fuyh-betwo) but was not addressed so far by Google.

Help and/or input is certainly appreciated.

**Comment by gqzc** [comment]
Why add UA after the browser shortcut to display the modified UA correctly?

**Comment by neroux** [comment]
Could you rephrase the question? I am not quite sure what you mean by "add after browser shortcut".

**Comment by gqzc** [comment]
For example   
![qq 20180814190743](https://user-images.githubusercontent.com/42342148/44088384-7fc1b692-9ff5-11e8-959f-4223a7d98794.jpg)
D:\chrome.exe --user-agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36"

**Comment by neroux** [comment]
This is a Chrome command line parameter to which extensions have no access.

**Comment by gqzc** [comment]
I see. Thank you.

**Comment by gqzc** [comment]
How does this extension do iframe spoofing?
https://chrome.google.com/webstore/detail/user-agent-switcher-and-m/bhchdcejhohfmigjafbampogmaanbfkg
![2](https://user-images.githubusercontent.com/42342148/50398964-c447e880-07b6-11e9-9e6f-7fc95d20b407.jpg)
![1](https://user-images.githubusercontent.com/42342148/50398968-cc078d00-07b6-11e9-9f2f-f68c353d681b.jpg)

**Comment by gqzc** [comment]
@neroux Can we use this scheme in random-user-agent?

**Comment by neroux** [comment]
Possibly. It would require further analysis but - as mentioned in my tag to @tarampampam - I am pretty confident the same approach could be used for RUA.

**Comment by gqzc** [comment]
I'm really looking forward to it.

**Comment by neroux** [comment]
Unfortunately I cant promise how quickly I'll get to have a look at a potential implementation. So unless someone else steps up, no guarantees on how soon that might be available I am afraid.

**Comment by neroux** [comment]
I managed to analyse it a bit better and it appears the mentioned extension manages to change the iframe user agent only on Chrome. Firefox still returns the actual user agent.

On Chrome, however, the fix might be easier than originally assumed and the manifest flag added with 44b7f67 might do the trick.

**Comment by gqzc** [comment]
Thanks for your works

**Comment by gqzc** [comment]
Firefox, you can read this post, maybe it will help you.
https://gitlab.com/ntninja/user-agent-switcher/issues/21

**Comment by gqzc** [comment]
Firefox doesn't matter. I'm looking forward to Chrome. Take your time

**Comment by neroux** [comment]
Personally I believe Firefox should matter a lot more than Chrome, but thats an ideological thing :)
* https://arstechnica.com/gadgets/2018/12/the-web-now-belongs-to-google-and-that-should-worry-us-all/
* https://haschromefullytakenoveryet.com/

As for the Chrome version, I believe yesterday's commit should "fix" that for Chrome but it was rather a hotfix than a thoroughly tested one, to be honest. @tarampampam just needs to publish a new version now :).

**Comment by neroux** [comment]
> the extension in question does seem to be able to override the user agent correctly on Firefox.

Hmm, earlier it seemed that way, now I cant reproduce that any longer and the extension in question does not manage to override the user agent in a Firefox/iframe context either.

@gqzc, can you confirm whether it actually works for you on Firefox?

**Comment by gqzc** [comment]
I use Chrome most of the time, Firefox only occasionally, so I don't think Firefox matters to me.

**Comment by gqzc** [comment]
@neroux Iframe can spoofing successfully, but why can't it spoofing on this website?
http://www.runoob.com/try/try.php?filename=try_nav_all
![123](https://user-images.githubusercontent.com/42342148/50947526-bc3bb880-14d9-11e9-937c-8f99c827bf6c.jpg)

**Comment by neroux** [comment]
@gqzc, I am afraid I am not quite sure about your question. It appears to be the identical you posed in #67, which actually led here.

**Comment by gqzc** [comment]
thank you all the same

**Comment by neroux** [comment]
@gqzc, version 2.2.8 which was just released should address this issue. Could you verify it?

**Comment by neroux** [comment]
I am afraid it appears as if 44b7f67 did not fix the issue on Chrome.

If anyone is up to fix this issue a pull request would be definitely appreciated.

**Comment by gqzc** [comment]
@neroux The new version solves this problem, but the problem of #67 still exists.
 Thanks for your works！

**Comment by BigCod** [comment]
https://github.com/kkapsner/CanvasBlocker

Might be able to help you with this issue as the extension manages to sucessfully protect the navigator api.

![image](https://user-images.githubusercontent.com/50107270/60768506-6163d200-a0b4-11e9-8ac0-ca1613220fca.png)

By using it I have managed to make this website believe I am using Firefox on linux, by spoofing the navigator api's output using the extension.

![image](https://user-images.githubusercontent.com/50107270/60768519-7ccedd00-a0b4-11e9-8ac4-e24d191b334e.png)

Just for reference I am actually using Firefox on windows.

Website used: http://f.vision

**Comment by neroux** [comment]
Thanks, @BigCod. It would seem they addressed this particular issue by introducing a dedicated frame handler, https://github.com/kkapsner/CanvasBlocker/blob/85320924fadb95b4939a5faf6869b6baee35b8b3/manifest.json#L63-L65

This probably could be adapted for RUA as well, might require some major code changes though. If you'd like to have a look at this, please submit a PR whenever convenient for you. Also tagging @tarampampam
